File: src/main/java/io/protobase/f7/formulas/math/SIN.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class SIN extends AbstractFormula {
  public static FormulaName NAME = FormulaName.SIN;
  public static SIN SELF = new SIN();

  public SIN() {
    super();
  }

  public SIN(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    return Math.sin(Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[0]))));
  }
}

File: src/main/js/formulas/math/SIN.ts
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class SIN extends AbstractFormula {
  static SELF: SIN = new SIN();
  NAME = FormulaName.SIN;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const value = Converters.toNumber(Converters.first(this.collateralLookup(origin, values[0])));
    return Math.sin(value);
  }
}
